src/sentry/static/sentry/app/views/settings/organizationDeveloperSettings/sentryApplicationRowButtons.tsx

```diff
@@ -0,0 +1,223 @@
+import Button from 'app/components/button';
+import ConfirmDelete from 'app/components/confirmDelete';
+import Confirm from 'app/components/confirm';
+import space from 'app/styles/space';
+import {t} from 'app/locale';
+import {Organization, SentryApp, SentryAppInstallation} from 'app/types';
+
+type Props = {
+  organization: Organization;
+  app: SentryApp;
+  install?: SentryAppInstallation;
+
+  isOnIntegrationPage: boolean;
+
+  onClickInstall?: () => void;
+  onClickUninstall?: (install: SentryAppInstallation) => void;
+
+  onClickRemove?: (app: SentryApp) => void;
+  onClickPublish?: () => void;
+};
+
+const SentryApplicationRowButtons = ({
+  organization,
+  app,
+  install,
+  isOnIntegrationPage,
+  onClickInstall,
+  onClickUninstall,
+
+  onClickRemove,
+  onClickPublish,
+}: Props) => {
+  const isInternal = app.status === 'internal';
+
+  // On the Integrations page, we show the install/uninstall buttons
+  if (isOnIntegrationPage) {
+    //no installation buttons to show if internal
+    if (isInternal) {
+      return null;
+    }
+    //if installed, render the uninstall button and if installed, render install button
+    return !!install ? (
+      <UninstallButton install={install} app={app} onClickUninstall={onClickUninstall} />
+    ) : (
+      <InstallButton onClickInstall={onClickInstall} />
+    );
+  }
+
+  // Otherwise, we are on the Developer Settings page and we want to show the action buttons
+  return (
+    <Access access={['org:admin']}>
+      {({hasAccess}) => {
+        let disablePublishReason = '';
+        let disableDeleteReason = '';
+
+        // Publish & Delete buttons will always be disabled if the app is published
+        if (app.status === 'published') {
+          disablePublishReason = t('Published integrations cannot be re-published.');
+          disableDeleteReason = t('Published integrations cannot be removed.');
+        } else if (!hasAccess) {
+          disablePublishReason = t(
+            'Organization owner permissions are required for this action.'
+          );
+          disableDeleteReason = t(
+            'Organization owner permissions are required for this action.'
+          );
+        }
+
+        return (
+          <ActionButtons
+            org={organization}
+            app={app}
+            // we only want to show the dashboard link on developer settings page
+            showDashboard={!isOnIntegrationPage}
+            showPublish={!isInternal}
+            showDelete
+            onPublish={onClickPublish}
+            onDelete={onClickRemove}
+            disablePublishReason={disablePublishReason}
+            disableDeleteReason={disableDeleteReason}
+          />
+        );
+      }}
+    </Access>
+  );
+};
+
+type UninstallButtonProps = {
+  install: SentryAppInstallation;
+  app: SentryApp;
+  onClickUninstall?: (install: SentryAppInstallation) => void;
+};
+const UninstallButton = ({install, app, onClickUninstall}: UninstallButtonProps) => {
```

Comment:
@billyvg What are your thoughts about having all these components in one file?

Reply:
Hmm, I don't think there's a single correct answer... I think these are all acceptable:
* as is 
* you could argue that `Uninstall` and `InstallButton` could be even inlined (like I think it was before)
* the `if (isOnIntegrationPage)` block as a component, since it does something pretty different than when you're not on integration page


I think one thing that would help is moving `ActionButtons` (and related sub components) to its own file.

I would also consider this file/folder structure: change `sentryApplicationRow.tsx` to a folder and have an `index.tsx` and move these components into that folder (assuming these components are *only* used by `sentryApplicationRow`'. That way you have some sort of grouping and association of what components are used by others by looking at folders.

Reply:
@billyvg @nolachen I like the idea a new folder then putting these components into different files in the same folder

Reply:
Yup I like the new folder structure better, I'll do that